File: DailyChallenges/CPP/LeetCode_881.cpp
class Solution {
public:
    int numRescueBoats(vector<int>& people, int limit) {
        sort(people.begin(), people.end());
        int l = 0, r = people.size(), ans = 0;
        while (l < r){
            if (people[--r] + people[l] <= limit) l++;
            ans++;
        }
        return ans;
    }
};

File: DailyChallenges/Scala/LeetCode_881.scala
object Solution {
    def numRescueBoats(people: Array[Int], limit: Int): Int = {
        import scala.collection.mutable.Queue
        val q = Queue.from(people.sorted.reverse.toIterable) // sorted queue
        
        def count(c: Int): Int = if (q.isEmpty) c else {
            val first = q.dequeue
            if (q.nonEmpty && (q.last + first <= limit)) q.dropRightInPlace(1)
            count(c+1)
        }
        
        count(0)
    }
}
